pubsub/pubsub.go

```diff
@@ -534,3 +536,30 @@ func wrapError(ec errorCoder, err error) error {
 	}
 	return gcerr.New(ec.ErrorCode(err), err, 2, "pubsub")
 }
+
+// RunWorkerPool repeatedly fetches messages from the given subscription sub, in an
+// inverted worker pool (link) with up to maxWorkers. The callback handleMessage is
+// called on each message. If handleMessage return nil for its error, then the
+// message is acked. If sub.Receive or handleMessage returns a non-nil error then
+// the workerpool is terminated via context cancellation.
+func RunWorkerPool(ctx context.Context, maxWorkers int, sub *Subscription, handleMessage func(ctx context.Context, m *Message) error) error {
+	errc := make(chan error)
+	ctx, cancel := context.WithCancel(ctx)
+	workerpool.Run(ctx, maxWorkers, func(ctx context.Context) workerpool.Task {
+		m, err := sub.Receive(ctx)
+		if err != nil {
+			cancel()
+			errc <- fmt.Errorf("receiving message: %v", err)
+			return func(context.Context) {}
+		}
+		return func(ctx context.Context) {
+			if err := handleMessage(ctx, m); err != nil {
+				cancel()
+				errc <- fmt.Errorf("handling message: %v", err)
```

Comment:
Changed it to have a buffer with `max+1` since there could be up to `max` task goroutines and maybe also the goroutine running `workerpool.Run` trying to send on the channel.